src/Models/HydrostaticFreeSurfaceModels/implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,77 @@
+using Oceananigans.Solvers
+using Oceananigans.Operators
+
+struct ImplicitFreeSurfaceSolver{S}
+    solver   :: S
+end
+
+function ImplicitFreeSurfaceSolver(arch, template_field, 
+                                   vertically_integrated_lateral_face_areas;
+                                   Amatrix_operator=nothing, 
+                                   maxit=nothing, 
+                                   tol=nothing)
+       
+    ## This should probably end up in some operator and grid generic auxilliaries
+    ## NEED TO SORT OUT HALO FOR PERIODIC CASE
+    Ax_baro = vertically_integrated_lateral_face_areas.Ax
+    Ay_baro = vertically_integrated_lateral_face_areas.Ay
+    @inline Ax_∂xᶠᵃᵃ_baro(i, j, k, grid, c) = Ax_baro[i, j, 1] * ∂xᶠᵃᵃ(i, j, 1, grid, c)
+    @inline Ay_∂yᵃᶠᵃ_baro(i, j, k, grid, c) = Ay_baro[i, j, 1] * ∂yᵃᶠᵃ(i, j, 1, grid, c)
+    @inline function ∇²_baro(i, j, k, grid, c)
+       return  δxᶜᵃᵃ(i, j, 1, grid, Ax_∂xᶠᵃᵃ_baro, c) +
+               δyᵃᶜᵃ(i, j, 1, grid, Ay_∂yᵃᶠᵃ_baro, c)
+    end
+
+    @kernel function implicit_η!(grid, f, implicit_η_f)
+        i, j = @index(Global, NTuple)
+        ### Not sure what to call this
+        ### it is for left hand side operator in
+        ### ( ∇ʰ⋅H∇ʰ - 1/gΔt² ) ηⁿ⁺¹ = 1/(gΔt) ∇ʰH U̅ˢᵗᵃʳ - 1/(gΔt²) ηⁿ
```

Comment:
Why `Δt²` rather than `Δt`? Can you explain where the left most term comes from?

Reply:
I am about to add something in the docs. Since it took me a while to get it right. It comes from applying 2.84 here (  https://github.com/CliMA/Oceananigans.jl/blob/a211d9d0f0ddf3f66ca9fe4319f47aac1e9e6164/src/Solvers/preconditioned_conjugate_gradient_solver.jl#L173  ) to the equations here ( https://github.com/CliMA/Oceananigans.jl/blob/a211d9d0f0ddf3f66ca9fe4319f47aac1e9e6164/src/Solvers/preconditioned_conjugate_gradient_solver.jl#L173 ). 

But it isn't well documented, so I was going to add to MITgcm and Oceanigans docs.

Reply:
There could be typo in the comment too - I need to check. The discrete line below is what we end up solving!

Reply:
Hmm, the docs make sense. I'll work out the algebra just to confirm!

Reply:
Ok, the algebra makes sense. It might be a bit simpler if we don't divide by `g`? But that doesn't change the answer.

Reply:
We could change that. I think we then need to multiply by g in the del^2 part.  Maybe that makes more sense, especially when we get to time dependent tidal forcing!!!